mypyc/genops.py

```diff
@@ -1603,158 +1604,138 @@ def for_loop_helper(self, index: Lvalue, expr: Expression,
+                                expr: Expression,
+                                body_block: BasicBlock,
+                                normal_loop_exit: BasicBlock,
+                                line: int,
+                                nested: bool = False) -> ForGenerator:
+        """Return helper object for generating a for loop over an iterable."""
+
+        if is_list_rprimitive(self.node_type(expr)):
+            # Special case "for x in <list>".
+            expr_reg = self.accept(expr)
             target_list_type = self.types[expr]
             assert isinstance(target_list_type, Instance)
             target_type = self.type_to_rtype(target_list_type.args[0])
 
-            value_box = self.translate_special_method_call(
-                self.read(expr_target, line), '__getitem__',
-                [self.read(index_target, line)], None, line)
-            assert value_box
-
-            self.assign(self.get_assignment_target(index),
-                        self.unbox_or_cast(value_box, target_type, line), line)
-
-            body_insts()
-
-            self.goto_and_activate(increment_block)
-            self.assign(index_target, self.primitive_op(
-                unsafe_short_add,
-                [self.read(index_target, line), self.add(LoadInt(1))], line), line)
-            self.goto(condition_block)
-
-            self.pop_loop_stack()
+            for_list = ForList(self, index, body_block, normal_loop_exit, line)
+            for_list.init(expr_reg, target_type)
+            return for_list
 
+        if (isinstance(expr, CallExpr)
+                and isinstance(expr.callee, RefExpr)):
+            if (expr.callee.fullname == 'builtins.range'
+                    and len(expr.args) <= 2
+                    and set(expr.arg_kinds) == {ARG_POS}):
+                # Special case "for x in range(...)".
+                # Only support 1 and 2 arg forms for now.
+                if len(expr.args) == 1:
+                    start_reg = self.add(LoadInt(0))
+                    end_reg = self.accept(expr.args[0])
+                else:
+                    start_reg = self.accept(expr.args[0])
+                    end_reg = self.accept(expr.args[1])
+
+                for_range = ForRange(self, index, body_block, normal_loop_exit, line)
+                for_range.init(start_reg, end_reg)
+                return for_range
+
+            elif (expr.callee.fullname == 'builtins.enumerate'
+                    and len(expr.args) == 1
+                    and expr.arg_kinds == [ARG_POS]
+                    and isinstance(index, TupleExpr)
+                    and len(index.items) == 2):
+                # Special case "for i, x in enumerate(y)".
+                lvalue1 = index.items[0]
+                lvalue2 = index.items[1]
+                if nested:
+                    cleanup_block = normal_loop_exit
+                else:
+                    cleanup_block = BasicBlock()
```

Comment:
This block (figuring out `cleanup_block`) is repeated three times and took me some puzzling to figure out. Is there a way to make it more clear/more factored out?